### PR TITLE
[PRM-2675] Implement new job to handle files not processed by SDES

### PR DESCRIPTION
**Pull request by JBlacklaws**

### CODE REVIEW

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
Could you fix the tests please?

**Comment by JBlacklaws** [comment]
> Could you fix the tests please?

Whoops! Thought I'd updated the application.conf already

File: app/config/AppConfig.scala [hunk 1]
```diff
@@ -42,5 +42,7 @@ class AppConfig @Inject()(config: Configuration, servicesConfig: ServicesConfig)
 
   lazy val minutesUntilNextAttemptOnCallbackFailure: Int = config.get[Int]("notifications.minutesUntilRetryOnCallbackFailure")
 
+  lazy val configurableTimeMinutes: Int = config.get[Int]("notifications.configurableTimeMinutes")
```
Comment (LewisMitchell) [inline app/config/AppConfig.scala:45]: Could this be named more descriptively?

File: app/services/NotProcessedFilesService.scala [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  //scalastyle:off
+  override def invoke: Future[Either[JobFailed, String]] = {
+    tryLock {
+      logger.info(s"[$jobName][invoke] - Job started")
+      for {
+        filesReceived <- fileNotificationRepository.getFilesReceivedBySDES()
+        filteredFiles = {
+          logger.info(s"[NotProcessedFilesService][invoke] - Number of files received by SDES: ${filesReceived.size}")
```
Comment (LewisMitchell) [inline app/services/NotProcessedFilesService.scala:59]: This seems a bit misleading, could we change it to be something like:
```suggestion
          logger.info(s"[NotProcessedFilesService][invoke] - Number of records in $FileReceivedInSDES state: ${filesReceived.size}")
```

File: app/services/NotProcessedFilesService.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        sequenceOfResults <- Future.sequence(filteredFiles.map {
+          logger.info(s"[NotProcessedFilesService][invoke] - Number of filtered files: ${filteredFiles.size}")
+          wrapper => {
+            PagerDutyHelper.log("invoke", NOTIFICATION_SET_TO_NOT_PROCESSED_PENDING_RETRY)
+            logger.info(s"[NotProcessedFilesService][invoke] - Updating notification (reference: ${wrapper.reference}) to NOT_PROCESSED_PENDING_RETRY")
+            fileNotificationRepository.updateFileNotification(wrapper.reference, RecordStatusEnum.NOT_PROCESSED_PENDING_RETRY).map(_ => true)
+          }.recover {
+            case e => {
+              PagerDutyHelper.log("invoke", UNKNOWN_EXCEPTION_FROM_SDES)
```
Comment (LewisMitchell) [inline app/services/NotProcessedFilesService.scala:72]: This PgaerDuty doesn't look right, we aren't receiving anything from SDES at this stage

File: app/services/NotProcessedFilesService.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        filesReceived <- fileNotificationRepository.getFilesReceivedBySDES()
+        filteredFiles = {
+          logger.info(s"[NotProcessedFilesService][invoke] - Number of files received by SDES: ${filesReceived.size}")
+          filesReceived.filter(notification => {
+            notification.nextAttemptAt.plusMinutes(appConfig.configurableTimeMinutes).isBefore(timeMachine.now)
+          })
+        }
+        sequenceOfResults <- Future.sequence(filteredFiles.map {
+          logger.info(s"[NotProcessedFilesService][invoke] - Number of filtered files: ${filteredFiles.size}")
+          wrapper => {
```
Comment (LewisMitchell) [inline app/services/NotProcessedFilesService.scala:66]:
```suggestion
          notification => {
```

File: app/services/NotProcessedFilesService.scala [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            PagerDutyHelper.log("invoke", NOTIFICATION_SET_TO_NOT_PROCESSED_PENDING_RETRY)
+            logger.info(s"[NotProcessedFilesService][invoke] - Updating notification (reference: ${wrapper.reference}) to NOT_PROCESSED_PENDING_RETRY")
+            fileNotificationRepository.updateFileNotification(wrapper.reference, RecordStatusEnum.NOT_PROCESSED_PENDING_RETRY).map(_ => true)
+          }.recover {
+            case e => {
+              PagerDutyHelper.log("invoke", UNKNOWN_EXCEPTION_FROM_SDES)
+              logger.error(s"[NotProcessedFilesService][invoke] - Exception occurred processing notifications - message: $e")
```
Comment (LewisMitchell) [inline app/services/NotProcessedFilesService.scala:73]: Do we have access to `notification` / `wrapper` here? If so, could we log the reference field to allow traceability?

File: app/utils/PagerDutyHelper.scala [hunk 6]
```diff
@@ -31,6 +31,7 @@ object PagerDutyHelper {
     final val FAILED_TO_PARSE_REQUEST_TO_MODEL = Value
     final val FAILED_TO_INSERT_FILE_NOTIFICATION = Value
     final val NOTIFICATION_SET_TO_PERMANENT_FAILURE = Value
+    final val NOTIFICATION_SET_TO_NOT_PROCESSED_PENDING_RETRY = Value
```
Comment (LewisMitchell) [inline app/utils/PagerDutyHelper.scala:34]: Need an alert config PR otherwise we'd never know this was being thrown

File: conf/application.conf [hunk 7]
```diff
@@ -146,6 +146,13 @@ schedules {
     enabled = false
     mongoLockTimeout = 7200 # 2 hours in seconds used to release the mongo lock in case the job fails with exception and doesn't unlock
   }
+
+  NotProcessedFilesService {
+    description = "Job to set files recieved SDES that haven't updated to NOT_PROCESSED_PENDING_RETRY"
```
Comment (LewisMitchell) [inline conf/application.conf:151]: Bit more description needed:
```suggestion
    description = "Job to set retry file notifications where FileProcessed has not been received by SDES"
```

File: conf/application.conf [hunk 8]
```diff
@@ -160,4 +167,5 @@ feature-switch {
 
 notifications {
   minutesUntilRetryOnCallbackFailure = 30
+  configurableTimeMinutes = 60
```
Comment (LewisMitchell) [inline conf/application.conf:170]: Could we be specific about what this does?

File: it/services/NotProcessedFilesServiceISpec.scala [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    numberOfAttempts = 1,
+    createdAt = LocalDateTime.of(2020,1,1,1,1),
+    updatedAt = LocalDateTime.of(2020,2,2,2,2),
+    nextAttemptAt = LocalDateTime.of(2020,3,3,3,3),
+    notification = sampleNotification
+  )
+
+  val pendingNotifications: Seq[SDESNotificationRecord] = Seq(
+    notificationRecord,
+    notificationRecord.copy(reference = "ref1", updatedAt = dateTimeOfNow, nextAttemptAt = LocalDateTime.of(2020,3,3,3,3)),
```
Comment (LewisMitchell) [inline it/services/NotProcessedFilesServiceISpec.scala:61]:
```suggestion
    notificationRecord.copy(reference = "ref1", updatedAt = dateTimeOfNow, nextAttemptAt = LocalDateTime.of(2020, 3, 3, 3, 3)),
```

File: it/services/NotProcessedFilesServiceISpec.scala [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    notificationRecord.copy(reference = "ref2", nextAttemptAt = dateTimeOfNow.minusMinutes(2), status = RecordStatusEnum.FILE_RECEIVED_IN_SDES),
+    notificationRecord.copy(reference = "ref3", nextAttemptAt = LocalDateTime.of(2020,3,3,3,3), status = RecordStatusEnum.FILE_RECEIVED_IN_SDES),
```
Comment (LewisMitchell) [inline it/services/NotProcessedFilesServiceISpec.scala:63]:
```suggestion
    notificationRecord.copy(reference = "ref3", nextAttemptAt = LocalDateTime.of(2020, 3, 3, 3, 3), status = RecordStatusEnum.FILE_RECEIVED_IN_SDES),
```

File: it/services/NotProcessedFilesServiceISpec.scala [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    notificationRecord.copy(reference = "ref4", nextAttemptAt = LocalDateTime.of(2020,3,3,3,3), status = RecordStatusEnum.NOT_PROCESSED_PENDING_RETRY)
```
Comment (LewisMitchell) [inline it/services/NotProcessedFilesServiceISpec.scala:64]:
```suggestion
    notificationRecord.copy(reference = "ref4", nextAttemptAt = LocalDateTime.of(2020, 3, 3, 3, 3), status = RecordStatusEnum.NOT_PROCESSED_PENDING_RETRY)
```

File: it/services/NotProcessedFilesServiceISpec.scala [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+      await(lockRepository.takeLock(service.lockKeeper.lockId, randomServerId, releaseDuration))
+      await(lockRepository.collection.countDocuments().toFuture()) shouldBe 1
+
+      await(service.invoke).getOrElse("fail") shouldBe s"${service.jobName} - JobAlreadyRunning"
+      await(lockRepository.collection.countDocuments().toFuture()) shouldBe 1
+    }
+  }
+
+  "invoke" should {
+    "run the job successfully if there is not notifications" in new Setup {
```
Comment (LewisMitchell) [inline it/services/NotProcessedFilesServiceISpec.scala:81]: Sounds like there is a dependency on there being notifications:
```suggestion
    "run the job successfully even if there is no notification" in new Setup {
```

File: it/services/NotProcessedFilesServiceISpec.scala [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+      val result = await(service.invoke)
+      result.isRight shouldBe true
+      result.getOrElse("fail") shouldBe "Processed all notifications"
+    }
+
+    "process the notifications and return Right is they all succeed - only process notifications where nextAttempt < now" in new Setup {
```
Comment (LewisMitchell) [inline it/services/NotProcessedFilesServiceISpec.scala:87]: Need to specify the logic about adding a set amount of time to the `updatedAt` timestamp before checking against the current date/time

File: it/services/NotProcessedFilesServiceISpec.scala [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+      SDESStub.successfulStubResponse()
+      await(notificationRepo.insertFileNotifications(pendingNotifications))
+      val result: Either[ScheduleStatus.JobFailed, String] = await(service.invoke)
+      result.isRight shouldBe true
+      result.getOrElse("fail") shouldBe "Processed all notifications"
+      val notificationsInRepo: Seq[SDESNotificationRecord] = await(notificationRepo.collection.find(Document()).toFuture())
+      notificationsInRepo.find(_.reference == "ref1").get.status shouldBe RecordStatusEnum.PENDING
+      notificationsInRepo.find(_.reference == "ref2").get.status shouldBe RecordStatusEnum.FILE_RECEIVED_IN_SDES
+      notificationsInRepo.find(_.reference == "ref3").get.status shouldBe RecordStatusEnum.NOT_PROCESSED_PENDING_RETRY
+      notificationsInRepo.find(_.reference == "ref3").get.updatedAt.isAfter(LocalDateTime.of(2020,2,2,2,2)) shouldBe true
```
Comment (LewisMitchell) [inline it/services/NotProcessedFilesServiceISpec.scala:97]:
```suggestion
      notificationsInRepo.find(_.reference == "ref3").get.updatedAt.isAfter(LocalDateTime.of(2020, 2, 2, 2, 2)) shouldBe true
```

File: test/services/MonitoringJobServiceSpec.scala [hunk 15]
```diff
@@ -121,7 +121,7 @@ class MonitoringJobServiceSpec extends SpecBase with LogCapturing {
     }
 
     s"return $Left ${MongoLockResponses.UnknownException} if lock returns exception, release lock is still called and succeeds" in new Setup {
-      val expectingResult = Future.successful(Right(Seq.empty))
+      val expectingResult: Future[Right[Nothing, Seq[Nothing]]] = Future.successful(Right(Seq.empty))
```
Comment (LewisMitchell) [inline test/services/MonitoringJobServiceSpec.scala:124]: I'm not sure what this is for?

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+import utils.PagerDutyHelper.PagerDutyKeys
+import utils.{LogCapturing, TimeMachine}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.concurrent.duration.{Duration, DurationInt}
+
+class NotProcessedFilesServiceSpec extends SpecBase with LogCapturing {
+  val mockLockRepository: MongoLockRepository = mock(classOf[MongoLockRepository])
+  val mockSDESConnector: SDESConnector = mock(classOf[SDESConnector])
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:46]: Not needed, we're not calling SDES in this job

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+    notificationRecord.copy(reference= "ref2", nextAttemptAt = mockDateTime.minusSeconds(1), status = RecordStatusEnum.FILE_RECEIVED_IN_SDES),
+    notificationRecord.copy(reference = "ref3", nextAttemptAt = mockDateTime.plusHours(1), status = RecordStatusEnum.FILE_RECEIVED_IN_SDES)
+  )
+
+  val pendingNotificationTwo = Seq(
+    notificationRecord,
+    notificationRecord.copy(reference= "ref2", nextAttemptAt = mockDateTime.minusSeconds(1)),
+    notificationRecord.copy(reference= "ref3", nextAttemptAt = mockDateTime.minusSeconds(3))
+  )
+
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:92]: What is the difference between this and `pendingNotifications` other than the nextAttemptAt?

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+  "invoke" should {
+    "run the job successfully if there are no relevant notifications" in new Setup {
+      when(mockFileNotificationRepository.getPendingNotifications()).thenReturn(Future.successful(Seq.empty))
+      when(mockFileNotificationRepository.getFilesReceivedBySDES()).thenReturn(Future.successful(Seq.empty))
+      val result: Either[ScheduleStatus.JobFailed, String] = await(service.invoke)
+      result.isRight shouldBe true
+      result.getOrElse("fail") shouldBe "Processed all notifications"
+    }
+
+    "process the notifications and return Right if they all succeed - only process if nextAttemptAt is < now" in new Setup {
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:116]: This needs to be updated to reflect the `plus 60 minutes` logic

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+      when(mockLockRepository.takeLock(Matchers.eq(mongoLockId), Matchers.any(), Matchers.eq(releaseDuration)))
+        .thenReturn(Future.successful(true))
+      when(mockLockRepository.releaseLock(Matchers.eq(mongoLockId), Matchers.any()))
+        .thenReturn(Future.successful(()))
+    }
+  }
+
+  "invoke" should {
+    "run the job successfully if there are no relevant notifications" in new Setup {
+      when(mockFileNotificationRepository.getPendingNotifications()).thenReturn(Future.successful(Seq.empty))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:109]: I don't think the service uses this method

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+      when(mockFileNotificationRepository.getFilesReceivedBySDES()).thenReturn(Future.successful(Seq.empty))
+      val result: Either[ScheduleStatus.JobFailed, String] = await(service.invoke)
+      result.isRight shouldBe true
+      result.getOrElse("fail") shouldBe "Processed all notifications"
+    }
+
+    "process the notifications and return Right if they all succeed - only process if nextAttemptAt is < now" in new Setup {
+      when(mockFileNotificationRepository.getPendingNotifications()).thenReturn(Future.successful(pendingNotifications))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:117]: I don't think the service uses this method

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+      ))
+      val result = await(service.invoke)
+      result.isRight shouldBe true
+      result.getOrElse("fail") shouldBe "Processed all notifications"
+      verify(mockFileNotificationRepository, times(1)).updateFileNotification(Matchers.eq("ref2"), Matchers.eq(RecordStatusEnum.NOT_PROCESSED_PENDING_RETRY))
+    }
+
+    "process the notifications and return Left if some fail" in new Setup {
+      val exception = new Exception("woopsy")
+      when(mockFileNotificationRepository.getPendingNotifications()).thenReturn(Future.successful(pendingNotificationTwo))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:130]: I don't think the service uses this method

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+            logs.exists(_.getMessage.contains(PagerDutyKeys.UNKNOWN_EXCEPTION_FROM_SDES))
+            logs.exists(_.getMessage.contains(PagerDutyKeys.FAILED_TO_PROCESS_FILE_NOTIFICATION))
+          }
+        }
+      }
+    }
+
+    "process the notifications and return Left if all fail" in new Setup {
+      val exception = new Exception("woopsy")
+      when(mockFileNotificationRepository.getPendingNotifications()).thenReturn(Future.successful(pendingNotificationTwo))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:150]: I don't think the service uses this method

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+      when(mockFileNotificationRepository.getFilesReceivedBySDES()).thenReturn(Future.successful(pendingNotificationTwo))
+      when(mockFileNotificationRepository.updateFileNotification(Matchers.any(), Matchers.any()))
+        .thenReturn(Future.failed(exception))
+        .thenReturn(Future.failed(exception))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:154]: .thenReturn will always return the same value no matter how many times its called, IIRC. You can remove the duplicate safely

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 24]
```diff
@@ -0,0 +1,230 @@
...
+      val result = await(service.invoke)
+      result.isLeft shouldBe true
+      result.left.getOrElse("fail") shouldBe FailedToProcessNotifications
+      withCaptureOfLoggingFrom(logger) {
+        logs => {
+          eventually {
+            logs.exists(_.getMessage.contains(PagerDutyKeys.UNKNOWN_EXCEPTION_FROM_SDES))
+            logs.exists(_.getMessage.contains(PagerDutyKeys.FAILED_TO_PROCESS_FILE_NOTIFICATION))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:162]: No assertions are made here - the test will always pass

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+        .thenReturn(Future.successful(notificationRecord.copy(reference = "ref2", status = RecordStatusEnum.NOT_PROCESSED_PENDING_RETRY, updatedAt = LocalDateTime.now())))
+        .thenReturn(Future.failed(exception))
+      val result = await(service.invoke)
+      result.isLeft shouldBe true
+      result.left.getOrElse("fail") shouldBe FailedToProcessNotifications
+      withCaptureOfLoggingFrom(logger) {
+        logs => {
+          eventually {
+            logs.exists(_.getMessage.contains(PagerDutyKeys.UNKNOWN_EXCEPTION_FROM_SDES))
+            logs.exists(_.getMessage.contains(PagerDutyKeys.FAILED_TO_PROCESS_FILE_NOTIFICATION))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:142]: No assertions are made here - the test will always pass

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 26]
```diff
@@ -0,0 +1,230 @@
...
+      val exception = new Exception("not again")
+      when(mockLockRepository.takeLock(Matchers.eq(mongoLockId), Matchers.any(), Matchers.eq(releaseDuration)))
+        .thenReturn(Future.failed(exception))
+      when(mockLockRepository.releaseLock(Matchers.eq(mongoLockId), Matchers.any()))
+        .thenReturn(Future.failed(exception))
+      withCaptureOfLoggingFrom(logger) { capturedLogEvents =>
+        await(service.tryLock(expectingResult)) shouldBe Left(MongoLockResponses.UnknownException(exception))
+        capturedLogEvents.exists(event => event.getLevel.levelStr == "INFO" && event.getMessage == s"[$jobName] Failed with exception") shouldBe true
+        eventually {
+          capturedLogEvents.exists(_.getMessage.contains(PagerDutyKeys.MONGO_LOCK_UNKNOWN_EXCEPTION))
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:223]: No assertion is made here - the tests will always pass. I can see this is the case in other files

File: app/services/NotProcessedFilesService.scala [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+    override val ttl: Duration = mongoLockTimeoutSeconds.seconds
+    override val lockRepository: LockRepository = lockRepositoryProvider
+  }
+
+  //scalastyle:off
+  override def invoke: Future[Either[JobFailed, String]] = {
+    tryLock {
+      logger.info(s"[$jobName][invoke] - Job started")
+      for {
+        filesReceived <- fileNotificationRepository.getFilesReceivedBySDES()
```
Comment (LewisMitchell) [inline app/services/NotProcessedFilesService.scala:57]: Could `fileReceived` be a bit more descriptive?

File: test/services/NotProcessedFilesServiceSpec.scala [hunk 28]
```diff
@@ -0,0 +1,230 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+
+import java.time.LocalDateTime
+
```
Comment (LewisMitchell) [inline test/services/NotProcessedFilesServiceSpec.scala:20]: Lots of unused imports here

**Review by LewisMitchell — DISMISSED** [review]

File: app/config/AppConfig.scala (hunk 1 above)
Comment (JBlacklaws) [reply app/config/AppConfig.scala:45]: Screenshot 2023-06-30 at 16 13 13
Not sure how better to describe it, should it be named in reference to the job?

Comment (LewisMitchell) [reply app/config/AppConfig.scala:45]: `configurableTimeMinutes` sounds quite generic as if it was used in a few places, ideally it should be named based on its use case

File: app/utils/PagerDutyHelper.scala (hunk 6 above)
Comment (JBlacklaws) [reply app/utils/PagerDutyHelper.scala:34]: https://github.com/hmrc/alert-config/pull/2919 Created

**Review by AndrewRButcher — APPROVED** [review]

**Pull request merged**